pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,394 @@
+		runtimeService:              runtimeService,
+		eventChannel:                eventChannel,
+		cache:                       cache,
+		genericPleg:                 genericPleg,
+		eventedPlegMaxStreamRetries: eventedPlegMaxStreamRetries,
+		relistDuration:              relistDuration,
+		clock:                       clock,
+	}
+}
+
+// Watch returns a channel from which the subscriber can receive PodLifecycleEvent events.
+func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
+	return e.eventChannel
+}
+
+// Relist relists all containers using GenericPLEG
+func (e *EventedPLEG) Relist() {
+	e.genericPleg.Relist()
+}
+
+// Start starts the Evented PLEG
+func (e *EventedPLEG) Start() {
+	e.runningMu.Lock()
+	defer e.runningMu.Unlock()
+	if isEventedPLEGInUse() {
+		return
+	}
+	setEventedPLEGUsage(true)
+	e.stopCh = make(chan struct{})
+	e.stopCacheUpdateCh = make(chan struct{})
+	go wait.Until(e.watchEventsChannel, 0, e.stopCh)
+	go wait.Until(e.updateGlobalCache, globalCacheUpdatePeriod, e.stopCacheUpdateCh)
+}
+
+// Stop stops the Evented PLEG
+func (e *EventedPLEG) Stop() {
+	e.runningMu.Lock()
+	defer e.runningMu.Unlock()
+	if !isEventedPLEGInUse() {
+		return
+	}
+	setEventedPLEGUsage(false)
+	close(e.stopCh)
+	close(e.stopCacheUpdateCh)
+}
+
+// In case the Evented PLEG experiences undetectable issues in the underlying
+// GRPC connection there is a remote chance the pod might get stuck in a
+// given state while it has progressed in its life cycle. This function will be
+// called periodically to update the global timestamp of the cache so that those
+// pods stuck at GetNewerThan in pod workers will get unstuck.
+func (e *EventedPLEG) updateGlobalCache() {
+	e.cache.UpdateTime(time.Now())
+}
+
+// Update the relisting period and threshold
+func (e *EventedPLEG) Update(relistDuration *RelistDuration) {
+	e.genericPleg.Update(relistDuration)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more
+	// than the relistThreshold. In case EventedPLEG is turned on,
+	// relistingPeriod and relistingThreshold are adjusted to higher
+	// values. So the health check of Generic PLEG should check
+	// the adjusted values of relistingPeriod and relistingThreshold.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) == cap(e.eventChannel) {
+		return false, fmt.Errorf("EventedPLEG: pleg event channel capacity is full with %v events", len(e.eventChannel))
+	}
+
+	timestamp := e.clock.Now()
+	metrics.PLEGLastSeen.Set(float64(timestamp.Unix()))
+	return true, nil
+}
+
+func (e *EventedPLEG) watchEventsChannel() {
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
+	defer close(containerEventsResponseCh)
+
+	// Get the container events from the runtime.
+	go func() {
+		numAttempts := 0
+		for {
+			if numAttempts >= e.eventedPlegMaxStreamRetries {
```

Comment:
I might be misunderstanding the logic here, but does this mean that after `eventedPlegMaxStreamRetries` we completely stop using evented pleg and fallback to generic PLEG?

Do we ever again re-attempt to restart evented pleg after the retry limit is reached? It seems like we should... for example take the case that you stop the container runtime for some time to perform some maintenance, during this time we could easily reach up to 5 retries (the current `eventedPlegMaxStreamRetries`). However after starting the container runtime, the connection could be established but it seems like the evented pleg would never re-start and re-connect? Maybe this needs some type of backoff logic instead?

Reply:
That's a good point. For now, we would require kubelet to be restarted in this scenario as we do not have a very deterministic way of retrying to re-establish the gRPC streaming connection with the runtime.

Either we have to consider a reasonable time period for retrying within this mechanism or have a way for runtime to inform kubelet that connection should be re-established. IMHO, we could consider this for beta. Thanks!